finesse/gui/hardware_set/hardware_sets_view.py

```diff
@@ -0,0 +1,145 @@
+
+
+class HardwareSetsControl(QGroupBox):
+    """A panel for choosing between hardware sets and (dis)connecting."""
+
+    def __init__(self) -> None:
+        """Create a new HardwareSetsControl."""
+        super().__init__("Hardware set")
+
+        self._connected_devices: set[OpenDeviceArgs] = set()
+        pub.subscribe(self._on_device_opened, "device.opening")
+        pub.subscribe(self._on_device_closed, "device.closed")
+
+        self._hardware_sets_combo = QComboBox()
+        self._hardware_sets_combo.setSizePolicy(
+            QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Minimum
+        )
+
+        # Populate combo box
+        for hw_set in load_builtin_hardware_sets():
+            self._add_hardware_set(hw_set)
+
+        # Remember which hardware set was in use the last time the program was run
+        last_selected = cast(str | None, settings.value("hardware_set/selected"))
+        if last_selected:
+            self._hardware_sets_combo.setCurrentText(last_selected)
+
+        self._hardware_sets_combo.currentIndexChanged.connect(
+            self._update_control_state
+        )
+
+        self._connect_btn = QPushButton("Connect")
+        self._connect_btn.setSizePolicy(
+            QSizePolicy.Policy.Minimum, QSizePolicy.Policy.Minimum
+        )
+        self._connect_btn.pressed.connect(self._on_connect_btn_pressed)
+        self._disconnect_btn = QPushButton("Disconnect all")
+        self._disconnect_btn.setSizePolicy(
+            QSizePolicy.Policy.Minimum, QSizePolicy.Policy.Minimum
+        )
+        self._disconnect_btn.pressed.connect(self._on_disconnect_btn_pressed)
+
+        # Enable/disable controls
+        self._update_control_state()
+
+        layout = QHBoxLayout()
+        layout.addWidget(self._hardware_sets_combo)
+        layout.addWidget(self._connect_btn)
+        layout.addWidget(self._disconnect_btn)
+        self.setLayout(layout)
+
+    def _add_hardware_set(self, hw_set: HardwareSet) -> None:
+        """Add a new hardware set to the combo box."""
+        labels = {
+            self._hardware_sets_combo.itemText(i)
+            for i in range(self._hardware_sets_combo.count())
+        }
+        if hw_set.name not in labels:
+            self._hardware_sets_combo.addItem(hw_set.name, hw_set)
+            return
+
+        # If there is already a hardware set by that name, append a number
+        i = 2
+        while True:
+            name = f"{hw_set.name} ({i})"
+            if name not in labels:
+                self._hardware_sets_combo.addItem(name, hw_set)
+                return
+            i += 1
+
+    @property
+    def current_hardware_set(self) -> frozenset[OpenDeviceArgs]:
+        """Return the currently selected hardware set."""
+        return self._hardware_sets_combo.currentData().devices
+
+    def _update_control_state(self) -> None:
+        # Enable the "Connect" button if there are any devices left to connect for this
+        # hardware set
```

Comment:
Make this a docstring?

Reply:
I've added a more generic docstring for the method.